### PR TITLE
8317538: Potential bottleneck in Provider::getService: specjvm2008::crypto.rsa have scalability issue for high vCPU numbers

### PR DESCRIPTION
**Pull request by IvaVladimir**
This patch remove access to the shared variable to fix scalability issue in the multithread environment. According to testing by the specjvm2008::crypto.rsa the one thread performance reduced for less than 1% while the score for the multithread run increased in ~2x. For the 2 socket system with Xeon 8480+ numbers looks as:
•	1 thread: 643.15 for original version vs 642.54 for patched one;
•	224 threads: 22446.19 for original vs 46147.41 for patched.
The RSABench microbenchmark reports no score changes for the 1 thread (average for all testcases) and 2.4% improvement for the 224 threads.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317538](https://bugs.openjdk.org/browse/JDK-8317538): Potential bottleneck in Provider::getService: specjvm2008::crypto.rsa have scalability issue for high vCPU numbers (**Bug** - P3)

### Reviewers
 * [Anthony Scarpino](https://openjdk.org/census#ascarpino) (@ascarpino - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21719/head:pull/21719` \
`$ git checkout pull/21719`

Update a local copy of the PR: \
`$ git checkout pull/21719` \
`$ git pull https://git.openjdk.org/jdk.git pull/21719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21719`

View PR using the GUI difftool: \
`$ git pr show -t 21719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21719.diff">https://git.openjdk.org/jdk/pull/21719.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21719#issuecomment-2438543890)
</details>

### CODE REVIEW

**Comment by ascarpino** [comment]
I have some concerns that removing `previousKey`, this will cause other benchmarks to suffer because a `SecretKey` instance is generated every time.  Have you tried using ThreadLocal with `previousKey`?  That may allow a similar performance gain, while allowing repeated use of the same key saving object allocations costs.

**Comment by IvaVladimir** [comment]
With the non-static field the reported score similar to the score with removed variable. I.e. the minimal changes 
-     private static volatile ServiceKey previousKey =
+    private transient ServiceKey previousKey =
gives a little bit smaller score for the 224 threads runs vs variable remove but scores close to run-to-run variation.
For the first glance, this update should have no negative impact for performance for other apps/ benchmarks.
Reported scores for current/removed var/instance var for 1 thread runs are 641.98/639.4/644.47 and for 224 threads are 21271.09/47841.5/42045.25.

File: src/java.base/share/classes/java/security/Provider.java [hunk 1]
```diff
@@ -1311,7 +1311,7 @@ public Service getService(String type, String algorithm) {
     // re-use will occur e.g. as the framework traverses the provider
     // list and queries each provider with the same values until it finds
     // a matching service
-    private static volatile ServiceKey previousKey =
+    private transient ServiceKey previousKey =
```
Comment (ascarpino) [inline src/java.base/share/classes/java/security/Provider.java:1314]: `transient` is for serialization and I don't think this should affect performance here.  Have you tried just `private ServiceKey ...`?

File: src/java.base/share/classes/java/security/Provider.java [hunk 2]
```diff
@@ -1311,7 +1311,7 @@ public Service getService(String type, String algorithm) {
...
                                             new ServiceKey("", "", false);
```
Comment (ascarpino) [inline src/java.base/share/classes/java/security/Provider.java:1315]: Please fix the formatting while you are making the change.  If a second line is still needed, it should be 4 whitespaces from the left margin.  The formatting was wrong before your change.  Thanks

**Comment by IvaVladimir** [comment]
The ServiceKey is not serializable type and without 'transient' modifier the build failed with warning: "[serial] non-transient instance field of a serializable class declared with a non-serializable type".
I believe it is better to skip serialization for the cached value than annotate it as expected warning.

**Comment by IvaVladimir** [comment]
If you think the annotation is better could you explain why?
The 'transient' modifier not affect performance and can be changed to '@suppresswarning'

**Comment by IvaVladimir** [comment]
Update patch to have ServiceKey as instance field.
The tier1 testing is OK. Reported scores for the 8280 CPU are:
original: 1 thread - 641.97; 112 threads - 19996.79 ops/m
patched: 1 thread - 644.21; 112 threads - 42840.73 ops/m

**Comment by ascarpino** [comment]
Changing serialization risks compatibility issues.  I strongly suggest looking into a ThreadLocal solution.

**Comment by IvaVladimir** [comment]
The ThreadLocal variable has similar compiler warning. With the field declared as:
===
    private ThreadLocal<ServiceKey> previousKey = new ThreadLocal<ServiceKey>();

Compiler for call 'previousKey.get();' reports: 
=== Output from failing command(s) repeated here ===
* For target jdk_modules_java.base__the.java.base_batch:
jdk/src/java.base/share/classes/java/security/Provider.java:1284: warning: [serial] non-transient instance field of a serializable class declared with a non-serializable type
        ServiceKey key = previousKey.get();

Could you add some words about scenario when 'transient' modifier leads to problem?
For now this field also not serialized.

**Comment by IvaVladimir** [comment]
Could you please check that 'transient' field with extra check is OK for your test repo?

**Comment by ascarpino** [comment]
ThreadLocal should be `static` as it is storing the per-thread `ServiceKey` variable.
You will want something like this:
```
private static final ThreadLocal<ServiceKey> previousKey =
        ThreadLocal.withInitial(() -> new ServiceKey("","", false));
```        
Then use previousKey.get() when reading the variable and set() for replacing it.

**Comment by IvaVladimir** [comment]
The tier1 tests are OK. The performance on 8480 with latest patch is:
original: 1 thread - 636.1; 112 threads - 15210.21 ops/m
patched: 1 thread - 641.56; 112 threads - 43410.4 ops/m

**Review by ascarpino — APPROVED** [review]
Thanks for working through all the issues to get this working correctly.  The current code looks good and has passed Tier 1-3 testing.

**Comment by ascarpino** [comment]
When are you going to integrate this?

**Comment by dwhite-intel** [comment]
Thanks @IvaVladimir and @ascarpino!

/sponsor

**Pull request closed without merging**